# PDS: Endterm - Practical part

### Date: February 26

### Time: 15:00 - 16:30


**Instructions:** 
* Write your full name and group number in the title of file;
* All answers should be filled in the notebook and then submitted to Moodle;
* Try to use fewer cells for compilation (for example, you can use print for several answers, instead of printing each answer on a separate cell)
* **IMPORTANT NOTE:** you will get points only for correct answers, not for trials or wrong answers. Your notebook should run if I will press **"Restart Kernel"** and then press **"Run All"**. I am going to check answers only after doing that. If the code do not compile, you will get zero. 

In [1]:
import pandas as pd
import numpy as np

## Problem 1 (1 point)

Create a function that takes non empty string as input, and returns Boolean output, which tells if a string is a polindrom. Polindrom is a string that has the same spelling from the left to the right and vice-versa. One symbol strings are also polyndroms. For example, 'aba' is a polindrom, while 'abc' is not a polindrom.

Use the following strings to check created function: 
1. 'a'
2. 'abcd'
3. 'abcdcba'
4. 'lol'
5. 'data'

In [15]:
def pal(st):
    return st == st[::-1]
    
for i in range(1, 6):
    s = input()
    print(pal(s))

a
True
abcd
False
abcdcba
True
lol
True
data
False


## Problem 2 (2 points)
Create a function that takes non empty string as input, and returns Boolean output, which tells if a string is an isogram. Isogram is a string that does not have repeating symbols. For example, 'abcd' is an isogram, while 'abca' is not an isogram.

Use the following strings to check created function: 
1. 'abcdcba'
2. 'abcdefg'
3. 'bowling'
4. 'lol'
5. 'data'

In [19]:
def iso(st):
    dc = {}
    for l in st:
        if l in dc:
            return False
        else:
            dc[l] = 1
        
    return True

for i in range(1, 6):
    s = input()
    print(iso(s))

abcdcba
False
abcdefg
True
bowling
True
lol
False
data
False


## Problem 3 (2 points)
You have an array of numbers and one number (k). Create a function, that moves this number k to the end the array, and then returns changed array. Function should perform all changes inplace, i.e. movements should take place inside the array without creating intermediary array, ignoring the order of numbers other than k. For example, array = [3,2,3,3,4,3,5,1,3,2] and k = 3, then after applying function you should get [2,4,5,1,2,3,3,3,3,3].

Use the following inputs to check created function:
1. [1,1,2] and 1
2. [1,2,2,1,2,1] and 2
3. [3,2,3,3,4,3,5,1,3,2] and 3


In [26]:
def push_back(arr, n):
    return sorted(arr, key=lambda x: x is n)    
    
for i in range(1, 4):
    array = input().replace(' ', '')
    num = input()
    print(push_back(array, num))

1 1 2
1
['2', '1', '1']
1 2 2 1 2 1
2
['1', '1', '1', '2', '2', '2']
3 2 3 3 4 3 5 1 3 2
3
['2', '4', '5', '1', '2', '3', '3', '3', '3', '3']


## Problem 4 (5 points)
For Coursera dataset (coursea_data.csv), do the tasks below.

Link to dataset: https://www.kaggle.com/siddharthm1698/coursera-course-dataset

### 1. Find three the most commonly appearing combination of certificate type and course difficulty. Show cerfiticate type, course difficulty and number of courses. (1 point)

In [51]:
df  = pd.read_csv('coursera_data.csv')

ds = df['course_Certificate_type'] + ' ' + df['course_difficulty']
ds.value_counts().head(3)

COURSE Beginner            282
SPECIALIZATION Beginner    196
COURSE Mixed               187
dtype: int64

### 2. Among 'Data Science' courses, find the course with the 2nd highest rating among Intermediate level courses. For obtained course, show its title, organization, and number of students enrolled. (1 point)

In [52]:
df

Unnamed: 0                                       course_title  \
0           134  (ISC)² Systems Security Certified Practitioner...   
1           743  A Crash Course in Causality:  Inferring Causal...   
2           874                     A Crash Course in Data Science   
3           413                            A Law Student's Toolkit   
4           635                A Life of Happiness and Fulfillment   
..          ...                                                ...   
886         236                         Программирование на Python   
887         883               Психолингвистика (Psycholinguistics)   
888         545       Разработка интерфейсов: вёрстка и JavaScript   
889         875                            Русский как иностранный   
890         163      Финансовые инструменты для частного инвестора   

                                   course_organization  \
0                                               (ISC)²   
1                           University of Pennsylvania   
2                             Johns Hopkins University   
3                                      Yale University   
4                            Indian School of Business   
..                                                 ...   
886                                      Mail.Ru Group   
887                  Saint Petersburg State University   
888                        E-Learning Development Fund   
889                  Saint Petersburg State University   
890  National Research University Higher School of ...   

    course_Certificate_type  course_rating course_difficulty  \
0            SPECIALIZATION            4.7          Beginner   
1                    COURSE            4.7      Intermediate   
2                    COURSE            4.5             Mixed   
3                    COURSE            4.7             Mixed   
4                    COURSE            4.8             Mixed   
..                      ...            ...               ...   
886          SPECIALIZATION            4.5      Intermediate   
887                  COURSE            4.8             Mixed   
888          SPECIALIZATION            4.5      Intermediate   
889          SPECIALIZATION            4.6      Intermediate   
890          SPECIALIZATION            4.7          Beginner   

    course_students_enrolled  
0                       5.3k  
1                        17k  
2                       130k  
3                        91k  
4                       320k  
..                       ...  
886                      52k  
887                      21k  
888                      30k  
889                     9.8k  
890                      38k  

[891 rows x 7 columns]

In [67]:
df1 = df.loc[df['course_difficulty'] == 'Intermediate']
df1 = df1.loc[df1['course_title'].str.contains('Data Science')]
df1.sort_values('course_rating', ascending = False)
print('so the second highest is: Applied Data Science with Python, specialized ')
df1.head(2)

so the second highest is: Applied Data Science with Python, specialized 


Unnamed: 0                      course_title     course_organization  \
55         637     Applied Data Science Capstone                     IBM   
56          26  Applied Data Science with Python  University of Michigan   

   course_Certificate_type  course_rating course_difficulty  \
55                  COURSE            4.7      Intermediate   
56          SPECIALIZATION            4.5      Intermediate   

   course_students_enrolled  
55                      42k  
56                     480k

### 3. For organizations from Ivy league (list provided below), find the one that has the highest presence in Coursera, i.e. the largest number of courses. For obtained organization, calculate average the rating for each of the certificate types. Show certificate type with mean rating.  (1 point)

In [7]:
# your code is here

### 4. Create a new column course_students_enrolled_new, which is the same as course_students_enrolled, but with numeric data type. To do that multiply the ones that ends to 'k' by 1000 and 'm' by 1000000. Show first 10 rows of the updated dataframe. (1 point)

*Hint: you might need to create a function, and then use apply() method.*

In [8]:
# your code is here

### 5. Assuming that all enrolled students will earn a certificate from the course, count how many certificates will be issued for each certificate type. Then order number of certificates issued in ascending order. And find share as a percentage (for example, 25.6) of each certificate type. Show result as a dataframe with columns certificate type, corresponding number and share. (1 point)

In [9]:
# your code is here

---

### Evaluation form (Keep it!)

| Question | Mark     | Comment   
|:-------:|:--------:|:----------------------
| 1       |   1/1    | Correct
| 2       |   2/2    | Correct
| 3       |   2/2    | Correct
| 4.1     |   1/1    | Correct
| 4.2     |   1/1    | Correct
| 4.3     |   1/1    | Correct
| 4.4     |   1/1    | Correct
| 4.5     |   1/1    | Correct
|**Total**|**10/10**  | 
